---
title: "Association mapping of the mouse gut microbiome"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Plotting
## Manhattan plots
Script run_manhattan(_histo).R will make the Manhattan plots using the pretty_manhattan.R script for each trait. 
Script run_overlay_manhattan.r takes the minimum P values of all the traits within one taxonomic level. 

## Region Plots
To make region plots run the run_pretty_region_plot.R script. It calls the New_pretty_plot_script.R. 
New_pretty_plot.R first calculates the confidence interval calculating the pairwise LD to the peak SNP and takes the positions from the two outer SNPs, that are in LD>0.9 with the peak SNP, as the borders. 
It then uses biomart to find the genes in the interval and plot this in a locuszoom-like manner. 
At the end, it will write a result file for the trait with the information about the significant regions.

## Effect plots
abundance_effect_plot.R takes a taxa, a taxonomic level, DNA or RNA and a marker name as input and will plot a boxplot of the abundance for each genotype and also prints the consensus genotype for musculus and domesticus mice. 

## Including Code

You can include R code in the document as follows:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
